### ISSUE TITLE
Growing weights

### ISSUE THREAD

**Issue by pefi9** [nicholas-leonard/rnn#3]
Hi,

thanks for great recurrent package. I'm still new to Torch, so the problem is probably in my usage. However, when I try to stack LSTM into multilayer network by using Sequencer, the weights are constantly growing. Without sequencer with only one recurrent layer it works fine.

network:
-- LSTM
```lua
    recurrent = nn.LSTM(inSize, hiddenSize1, rho)
    recurrent2 = nn.LSTM(hiddenSize1, hiddenSize2, rho)
    recurrent.scales = torch.Tensor(rho):fill(1)
    recurrent2.scales = torch.Tensor(rho):fill(1)
    linear = nn.Linear(hiddenSize2, outSize)

    model = nn.Sequential()
    sequencer = nn.Sequencer(recurrent)
    sequencer2 = nn.Sequencer(recurrent2)
    sequencer3 = nn.Sequencer(linear)
    model:add(sequencer)
    model:add(sequencer2)
    model:add(sequencer3)
```
-- TRAINING
```lua
for t = 1, (trainSize - batchSize) do

        -- load new sample
        local inputs, targets, gradOutputs = {}, {}, {}
        for step = 1, rho do
            local index = t + step
            inputs[step] = inputData:sub(1, inputData:size(1), index, index + batchSize - 1):transpose(1,2)
            targets[step] = outputData:sub(1, outputData:size(1), index, index + batchSize - 1):transpose(1, 2)

        end

        local outputs = model:forward(inputs)

        for step = 1, rho do
            local err = criterion:forward(outputs[step], targets[step])
            trainError = trainError + err
            gradOutputs[step] = criterion:backward(outputs[step], targets[step])
        end

        model:backward(inputs, gradOutputs)
        model:updateParameters(lrRNN)
        linear:zeroGradParameters()
    end
```

I have tried lowering the learning rate (1e-2, 1e-3, 1e-4), different size of scales (1, 1/rho), batch size (1,10) or rho (1,5,10,100). But none of these modification seemed to work. Any ideas?

**Comment by nicholas-leonard** [comment]
Hi,

Glad you like it. So I suspect the issue is in 
```lua 
gradOutputs[step] = criterion:backward(outputs[step], targets[step])
```
where it should be 
```lua
gradOutputs[step] =rnn.recursiveCopy(gradOutputs[step], criterion:backward(outputs[step], targets[step]))
```
Otherwise, `gradOutputs` will contain `rho` identical `gradOutput` Tensors. 

This shouldn't be a problem once I roll out SequencerCriterion.

**Comment by nicholas-leonard** [comment]
[SequencerCriterion](https://github.com/nicholas-leonard/rnn/blob/master/README.md#rnn.SequencerCriterion)

**Comment by pefi9** [comment]
Thanks for the reply. I`ve tried both solution, but without any progress.

**Comment by nicholas-leonard** [comment]
Darn. What dataset are you using? I would like to try it out on my end.

**Comment by pefi9** [comment]
It`s no general dataset. I`m trying to model time series. Thanks for help, I`ll do double check and optionally send you the set.

**Comment by AjayTalati** [comment]
Hey Nick and pefi9 - I'm working on modelling time series, using LSTM - it's not an easy ride that's for sure !!! 

The only things that are helping at the moment is printing output batch and epoch means, standard deviations, and max's / mins of the input and outputs of my LSTMs modules, and criterion's. 

It's helped to find a few mistake in my code - you know garage in = garbage out :(

**Comment by nicholas-leonard** [comment]
Yeah its full of obstacles. I really need to find the time to play with the LSTM.

**Comment by AjayTalati** [comment]
So what are the right things to monitor for a systems of LSTM modules that are unfolded from time `t=1,..., T`?  I guess it's the outputs of the `t=T` module, and the gradients of the `t=1` module, and you would use, `L2` norms as the stability measures for both the outputs and grad params?

**Comment by nicholas-leonard** [comment]
Yes. The norms of the gradients w.r.t. to parameters and outputs. You can monitor max and average norms over batches. Its messy, but it works :)

Nice plots.

**Comment by nicholas-leonard** [comment]
@pefi9 Try the latest version of rnn. I updated the Sequencer. Fixed a bug.

**Comment by AjayTalati** [comment]
Hey Nick, thanks for the great advice :+1: 

> The norms of the gradients w.r.t. to parameters and outputs. You can monitor max and average norms over batches. Its messy, but it works :)

I finally spotted what was going wrong - it was not an error in my understanding - which was drive me nuts - but it was that Oxford Msc course code I was using,

[model_utils.combine_all_parameters]( https://github.com/oxford-cs-ml-2015/practical6/blob/master/model_utils.lua )

I'm guessing this may well be unstable - does not work in some cases - at least for me it seems to give weird behaviour ??? 

When I wrapped all my modules, (reader, encoder, Q-sampler, decoder and writer). up into a big `gModule`, and used the standard, `:getParameters()`. Everything works better, and also the objective function that goes into the optimizer is a lot more tidy. I still don't get the same MNIST values as deepmind, but they're close - and at least it converges :+1: 

So basically I went nuts for 10 days, and had a sad Easter, because I used code written by someone else, which I did'nt really understand - won't be doing that again :-1:   

I'm just wondering if you've seen anything similar that you know is reliable - pretty sure it works?

**Comment by pefi9** [comment]
Sorry for delay, I`ve been offline for a while. 
@AjayTalati - Thanks for the advice, I`ll try to search deeper. 
@nicholas-leonard - I have created repository with the dataset if you are interested in ( https://github.com/pefi9/lstm_data_set ). I tried the last update, sadly I can`t confirm any change. As I wrote I`ll try to look at the gradients what might go wrong. Ii possible as well that I`m just using the package wrongly as lstm and torch are pretty new for me. I`ll let you know if I make any progress on this. 

Thanks a lot
Petr

**Comment by szagoruyko** [comment]
@pefi9 in your training loop you call ```linear:zeroGradParameters()``` whereas you should call ```model:zeroGradParameters()```, that makes your gradients and weights of LSTM layers to constantly grow.

**Comment by nicholas-leonard** [comment]
Thanks @szagoruyko

**Comment by pefi9** [comment]
@szagoruyko thanks a lot. That helped :-).

**Comment by AjayTalati** [comment]
@pefi9  -- a really simply and effective way to stop exploding weights, (for my cloned `nngraph` system), is just to train it for a very small number of time steps. Save it, rebuild the system and clone it for double the number of time steps, and then to copy the parameters into the shared parameter tensor, for the new (double length system). That's easy and it works :) 

I'm still trying to understand DP but when I finally move over to it, I'll try that!
